src/Package.cc

```
@@ -2942,3 +2945,49 @@ YCPValue PkgFunctions::CreateSolverTestCase(const YCPString &dir)
     return YCPBoolean(success);
 }
 
+/**
+ * Get a package object from a given repository
+ *
+ * @param YCPInteger   repo_id Repository ID (alias)
+ * @param YCPString    name    Package name
+ * @return zypp::Package::constPtr
+ */
+zypp::Package::constPtr PkgFunctions::packageFromRepo(const YCPInteger & repo_id, const YCPString & name) {
+  zypp::ResPool pool(zypp::getZYpp()->pool());
+  YRepo_Ptr repo = logFindRepository(repo_id->value());
+  if (!repo) return NULL;
+
+  /* maybe we should use std::find_if */
+  for_(it, pool.byIdentBegin<zypp::Package>(name->value()), pool.byIdentEnd<zypp::Package>(name->value())) {
+    if (repo->repoInfo().alias() == (*it)->repository().alias()) {
+      return zypp::asKind<zypp::Package>((*it).resolvable());
+    }
+  }
+  return NULL;
+}
+
+/**
+ * Provide a package using the \c zypp::repo::PackageProvider class
+ *
+ * @param YCPInteger   repo_id Repository ID (alias)
+ * @param YCPString    name    Package name
+ * @param YCPString    path    Path to save the file.
+ * @return YCPValue    If the package was found, it returns true;
+ *                     if it wasn't found, it returns false.
+ */
+YCPValue PkgFunctions::ProvidePackage(const YCPInteger & repo_id, const YCPString & name, const YCPString & path) {
+
+  zypp::Package::constPtr package = packageFromRepo(repo_id, name);
+  if (package == NULL) return YCPBoolean(false);
+
+  zypp::repo::RepoMediaAccess access;
+  zypp::repo::PackageProviderPolicy packageProviderPolicy;
+  zypp::repo::DeltaCandidates deltas;
+  zypp::repo::PackageProvider pkgProvider(access, package, deltas, packageProviderPolicy);
+  zypp::ManagedFile file(pkgProvider.providePackage());
```

What about the error handling? What happens if the package download fails? Does libzypp throw an exception? If yes we should handle it. (Maybe simply catch all exceptions and return `false` in that case, like [here](https://github.com/yast/yast-pkg-bindings/blob/42d2b703f87ec48f98b367df9536f8ba768feb01/src/Resolvable_Install.cc#L231).)